### PR TITLE
[UX-181] Introduce api to see queued and running items

### PR DESCRIPTION
**Pull request by imeredith**
This is a WIP.

### CODE REVIEW

**Comment by i386** [comment]
@imeredith don't forget the `README.md` update. Great work on getting this working though.

File: blueocean-plugin/src/main/java/io/jenkins/blueocean/service/embedded/rest/queue/BlueTaskQueueImpl.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public BlueRun getRun() {
+        return null;
+    }
+
+    public static Iterable<BlueTaskQueueImpl> getQueuedTasks() {
+        ImmutableList<Queue.Item> items = ImmutableList.copyOf(Jenkins.getInstance().getQueue().getItems());
+        return Iterables.transform(items, new Function<Queue.Item, BlueTaskQueueImpl>() {
```
Comment (tfennelly) [inline blueocean-plugin/src/main/java/io/jenkins/blueocean/service/embedded/rest/queue/BlueTaskQueueImpl.java]: So is this list effectively being cloned twice every time?

Why not just a simple `for` loop to create the `Collection` of `BlueTaskQueueImpl` instances? Aside from the fact that I had to read it twice to get what was happening, it seems like we're creating extra collections than are needed.

Comment (tfennelly) [inline blueocean-plugin/src/main/java/io/jenkins/blueocean/service/embedded/rest/queue/BlueTaskQueueImpl.java]: E.g.

```java
public static Iterable<BlueTaskQueueImpl> getQueuedTasks() {
    List<BlueTaskQueueImpl> blueItems = new ArrayList<BlueTaskQueueImpl>();
    for (Queue.Item item : Jenkins.getInstance().getQueue().getItems()) {
        blueItems.add(new BlueTaskQueueImpl(item));
    }
    return blueItems.iterable();
}
```

File: blueocean-plugin/src/main/java/io/jenkins/blueocean/service/embedded/rest/queue/BlueTaskRunImpl.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+public class BlueTaskRunImpl extends BlueTask {
+    private Run run;
+
+    public BlueTaskRunImpl(Run run) {
+        this.run = run;
+    }
+
+    public String getId() {
+        return Long.toString(run.getQueueId());
+    }
```
Comment (tfennelly) [inline blueocean-plugin/src/main/java/io/jenkins/blueocean/service/embedded/rest/queue/BlueTaskRunImpl.java]: I would thing we need a `getQueueId` and a `getId`.

File: blueocean-plugin/src/test/java/io/jenkins/blueocean/service/embedded/TasksTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            } else {
+                id =  p.getQueueItem().getId();
+            }
+
+            Map r = request().get("/tasks/" + id).build(Map.class);
+            Assert.assertEquals(r.get("id"), Long.toString(id));
+        }
+
+    }
+}
```
Comment (tfennelly) [inline blueocean-plugin/src/test/java/io/jenkins/blueocean/service/embedded/TasksTest.java]: A few comments on the tests would be good I think i.e. what the tests are testing for. They look like they are nearly doing the same thing but just some subtle differences that are not obvious for those not familiar with the code.

**Comment by tfennelly** [comment]
I think, in general, some Javadoc would be good on the new code.

I'm not sure if it's covered in this, but if I get a list of the runs that are currently executing, can I get the queueId of the original `Queue.Item` task that that run was created from i.e. correlate between them?

File: blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/queue/BlueTask.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    @Exported(name = TYPE)
+    public abstract TaskType getType();
+
+    @Exported(name = NAME)
+    public abstract String getName();
+
+    @Exported(name = RUN, inline = true)
+    public abstract BlueRun getRun();
+
+    public enum TaskType {
```
Comment (vivek) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/queue/BlueTask.java]: Whats the purpose of task type is? Enum type makes it specific to only these two types. For example, if my plugin wants to add it's own Task, how would the getType work? I think we should remove getType(), just the subclasses of BlueTask is what the type of task is.

File: blueocean-plugin/src/main/java/io/jenkins/blueocean/service/embedded/rest/AbstractRunImpl.java [hunk 5]
```diff
@@ -155,7 +160,7 @@ public BluePipelineNodeContainer getNodes() {
         return null; // default
     }
 
-    protected static BlueRun getBlueRun(Run r){
+    public static BlueRun getBlueRun(Run r){
```
Comment (vivek) [inline blueocean-plugin/src/main/java/io/jenkins/blueocean/service/embedded/rest/AbstractRunImpl.java]: Why this change in scope?

File: blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/queue/BlueTask.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    public abstract String getId();
+
+    @Exported(name = TYPE)
+    public abstract TaskType getType();
+
+    @Exported(name = NAME)
+    public abstract String getName();
+
+    @Exported(name = RUN, inline = true)
+    public abstract BlueRun getRun();
```
Comment (vivek) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/queue/BlueTask.java]: Not sure I understand, why is a task return run? Each of the exported methods should have comments to tell what its about, will be helpful to understand.

File: blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/queue/BlueTask.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    public static final String RUN = "run";
+
+    /**
+     * @returns Id of queue item. Needs to match up with {@link BlueRun#getQueueId()}.
+     */
+    @Exported(name = ID)
+    public abstract String getId();
+
+    @Exported(name = TYPE)
+    public abstract TaskType getType();
```
Comment (vivek) [inline blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/queue/BlueTask.java]: Not sure if we need it.
